### PR TITLE
Moving all allocas in entry blocks to the beginning of the block

### PR DESCRIPTION
**Pull request by adam-yang**

### CODE REVIEW

File: tools/clang/unittests/HLSL/CompilerTest.cpp [hunk 1]
```diff
@@ -2639,6 +2640,10 @@ TEST_F(CompilerTest, CodeGenSamples){
   CodeGenTestCheckBatchDir(L"Samples");
 }
 
+TEST_F(CompilerTest, AllocaAfterStaticInit) {
+  CodeGenTest(L"gv_memcpy_before_alloca.hlsl");
```
Comment (MrTrillian) [inline tools/clang/unittests/HLSL/CompilerTest.cpp]: We shouldn't be using `CodeGenTest`, it doesn't even run FileCheck, only asserts that the disassembly is non-empty. Also let's leverage batch tests. This should go into `batch/passes/sroa_hlsl`

**Comment by MrTrillian** [comment]
Should we not be finding places where we insert allocas at non-entry positions and fixing those instead? We have functions to find alloca insert positions and all, so if we're not using them it's probably a bug? It's a bit arbitrary to have SROA do this moving. What if other prior passes also depend on allocas being first?

File: lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp [hunk 2]
```diff
@@ -4325,6 +4327,10 @@ class SROA_Parameter_HLSL : public ModulePass {
     // bitcast ptr.
     MemcpySplitter::PatchMemCpyWithZeroIdxGEP(M);
 
+    // Move all Allocas to the front of the entry block, as the entry block might be
+    // split at some point, leaving some allocas stranded in a non-entry block.
+    FrontLoadAllAllocas(M);
```
Comment (MrTrillian) [inline lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp]: As this has no dependencies on SROA, can we make it a pass of its own?

**Comment by adam-yang** [comment]
No, the alloca insertion is correct as far as I can tell. We get into the situation here due to inlining functions.

**Comment by MrTrillian** [comment]
Oh, inlining, that makes sense.

**Comment by adam-yang** [comment]
Actually there is a codegen problem where alloca is inserted in the wrong place... I'm gonna do that instead. Shelving this for now.

**Pull request closed without merging**